mqtt/mqtt-edgehub/src/auth/authorization/edgehub.rs

```diff
@@ -141,67 +107,97 @@ where
+                //
+                // [edgehub] <- [actor = client (leaf or child edge)]
+                match self.identities_cache.get(&actor_id) {
+                    Some(identity) => identity
+                        .auth_chain()
+                        .map_or(false, |auth_chain| auth_chain.contains(&self.device_id)),
+                    None => false,
+                }
             }
-            Some(on_behalf_of_id) => self
-                .identities_cache
-                .get(&on_behalf_of_id)
-                .and_then(IdentityUpdate::auth_chain)
-                .map_or(false, |auth_chain| auth_chain.contains(client_id.as_str())),
-            None => {
-                // If there is no on_behalf_of_id, we are dealing with a legacy topic
-                // The client_id must still be in the identities cache
-                self.identities_cache.contains_key(client_id)
+            Some(actor_id) => {
+                // if actor != client, it means it is an on-behalf-of request.
+                // check that:
+                // - actor_id is in the auth chain for client_id (that client
+                //   making a request can actually do it on behalf of the actor)
+                // - check that actor_id is in the auth chain for current edgehub.
+                // - check that client_id is in the auth chain for current edgehub.
+                //
+                // [edgehub] <- [client (child edgehub)] <- [actor (grandchild)]
+
+                let parent_ok = match self.identities_cache.get(&actor_id) {
+                    Some(identity) => identity.auth_chain().map_or(false, |auth_chain| {
+                        auth_chain.contains(&client_id.as_str().replace("/$edgeHub", ""))
+                    }),
+                    None => false,
+                };
+
+                let actor_ok = match self.identities_cache.get(&actor_id) {
+                    Some(identity) => identity
+                        .auth_chain()
+                        .map_or(false, |auth_chain| auth_chain.contains(&self.device_id)),
+                    None => false,
+                };
+
+                let client_ok = match self.identities_cache.get(client_id) {
+                    Some(identity) => identity
+                        .auth_chain()
+                        .map_or(false, |auth_chain| auth_chain.contains(&self.device_id)),
+                    None => false,
+                };
+
+                parent_ok && actor_ok && client_ok
             }
+            // If there is no actor_id, we are dealing with a legacy topic/unknown format.
+            // Delegated to inner authorizer.
+            None => false,
         }
     }
 }
 
-fn get_on_behalf_of_id(topic: &str) -> Option<ClientId> {
-    // topics without the new topic format cannot have on_behalf_of_ids
-    if !topic.starts_with("$iothub/clients") {
-        return None;
+fn get_actor_id(topic: &str) -> Option<ClientId> {
```

Comment:
nit: `extract_client_id` or `extract_actor_id`. you added a notion of an actor which is not common in our codebase. I was a bit confused by this naming